src/model_repository_manager.cc

```diff
@@ -671,155 +653,237 @@ ModelRepositoryManager::LoadUnloadModels(
-        std::set<std::string> next_models;
-        // The intention here is to iteratively expand the collection of models
-        // to be polled for this load operation, if ensemble is involved,
-        // until the collection is self-contained.
-        // [FIXME] Every Poll() above add new entries in existing 'new_infos'
-        // (and 'added' etc.), so iterating on the whole structure means
-        // re-examination on certain entries.
-        std::set<std::string> temp_checked;
-        for (auto& info : new_infos) {
-          // if the model is not "checked", it is model polled at current
-          // iteration, check config for more models to be polled
-          const bool checked =
-              (checked_models.find(info.first.name_) != checked_models.end());
-          if (!checked) {
-            // Don't add the model name to 'checked_models' as there may be
-            // same name model polled but with different namespace
-            temp_checked.emplace(info.first.name_);
-            const auto& config = info.second->model_config_;
-            if (config.has_ensemble_scheduling()) {
-              for (const auto& step : config.ensemble_scheduling().step()) {
-                next_models.emplace(step.model_name());
-              }
-            }
-          }
+
+    // The models are in 'deleted' either when they are asked to be unloaded or
+    // they are not found / are duplicated across all model repositories.
+    // Explicitly remove them from infos.
+    for (const auto& name :
+         (unload_dependents ? deleted_dependents : deleted)) {
+      infos.map_.erase(name);
+    }
+
+    // At this point, it is committed to load/unload the models, so write the
+    // local copy into this object. The state of the models cannot be modified
+    // by another thread while they are locked. Thus, it is safe to release the
+    // mutex while actively loading/unloading models, but the mutex must be
+    // re-held when reading/updating any model info to avoid any corruption.
+    infos_ = infos;
+    global_map_ = global_map;
+    dependency_graph_.Assign(dependency_graph, &global_map_);
+
+    // Unlock the affected models.
+    // This serve 2 purposes:
+    // 1. The unlocked state will be writeback to the dependency graph owned by
+    //    this object, after the load/unload is completed.
+    // 2. It doubles as an indicator that all the locked models will be handled
+    //    by another thread, and this thread must not attempt to load/unload
+    //    those locked models.
+    dependency_graph.UnlockNodes(affected_models);
+  }
+
+  // Unload 'deleted' models.
+  for (const auto& name : (unload_dependents ? deleted_dependents : deleted)) {
+    model_life_cycle_->AsyncUnload(name);
+  }
+  // Load/Unload the models affected
+  const auto& load_status = LoadModelByDependency(&dependency_graph, &infos);
+
+  {
+    std::lock_guard<std::mutex> lock(mu_);
```

Comment:
why not use `unique_lock` instead if you need to lock / unlock mutex multiple time in the same scope.